File: c/library_2/git_lib/determine_incident_halffaces.c

#ifndef _INCLUDE_BASIC_
#define _INCLUDE_BASIC_

#include "basic_emxArray_fncs.c"

#endif

/*
 * DETERMINE_INCIDENT_HALFFACES Determine an incident half-faces.
 *  DETERMINE_INCIDENT_HALFFACES(ELEMS,OPPHFS,V2HF) Determines an
 *  incident half-faces of each vertex. Give higher priorities to border
 *  faces. The following explains the inputs and outputs.
 *
 *  Input:  ELEMS:  matrix of size mx4 storing element connectivity
 *          OPPHFS: matrix of size mx4 storing opposite half-faces
 *
 *  Output: V2HF:   array of size equal to number of vertices.
 *
 *  See also DETERMINE_INCIDENT_HALFEDGES, DETERMINE_INCIDENT_HALFVERTS
 */
void determine_incident_halffaces(const emxArray_int32_T *elems, const
  emxArray_int32_T *opphfs, emxArray_int32_T *v2hf)
{
  emxArray_int32_T *b_v2hf;
  int32_T nv;
  int32_T jj;
  int32_T loop_ub;
  emxArray_boolean_T *isborder;
  static const int8_T hf[24] = { 1, 1, 2, 3, 3, 2, 3, 1, 2, 4, 4, 4, 7, 5, 6, 7,
    5, 9, 10, 8, 6, 8, 9, 10 };

  static const int8_T v2f[12] = { 2, 1, 3, 4, 4, 3, 1, 2, 1, 2, 4, 3 };

  emxInit_int32_T(&b_v2hf, 1);

  /*  Table for local IDs of incident faces of each vertex. */
  /*  We use three bits for local-face ID. */
  /*  Construct a vertex to halfedge mapping. */
  nv = v2hf->size[0];
  jj = v2hf->size[0];
  v2hf->size[0] = v2hf->size[0];
  emxEnsureCapacity((emxArray__common *)v2hf, jj, (int32_T)sizeof(int32_T));
  loop_ub = v2hf->size[0] - 1;
  for (jj = 0; jj <= loop_ub; jj++) {
    v2hf->data[jj] = 0;
  }

  emxInit_boolean_T(&isborder, 1);

  /*  Table for local IDs of incident faces of each vertex. */
  /*  DETERMINE_BORDER_VERTICES_VOL Determine border vertices of a volume mesh. */
  /*  */
  /*  ISBORDER = DETERMINE_BORDER_VERTICES_VOL(NV,ELEMS,OPPHFS,ISBORDER) */
  /*  Determines border vertices of a volume mesh.  It supports both linear */
  /*     and quadratic elements. It returns bitmap of border vertices. For */
  /*     quadratic elements, vertices on edge and face centers are set to false, */
  /*     unless QUADRATIC is set to true at input. */
  /*  */
  /*  Example */
  /*  ISBORDER = DETERMINE_BORDER_VERTICES_VOL(NV,ELEMS) */
  /*  ISBORDER = DETERMINE_BORDER_VERTICES_VOL(NV,ELEMS,OPPHFS) */
  /*  ISBORDER = DETERMINE_BORDER_VERTICES_VOL(NV,ELEMS,OPPHFS,ISBORDER) */
  /*  ISBORDER = DETERMINE_BORDER_VERTICES_VOL(NV,ELEMS,OPPHFS,ISBORDER,LINEAR) */
  /*  */
  /*  See also DETERMINE_BORDER_VERTICES_CURV, DETERMINE_BORDER_VERTICES_SURF */
  jj = isborder->size[0];
  isborder->size[0] = nv;
  emxEnsureCapacity((emxArray__common *)isborder, jj, (int32_T)sizeof(boolean_T));
  loop_ub = nv - 1;
  for (jj = 0; jj <= loop_ub; jj++) {
    isborder->data[jj] = FALSE;
  }

  /*  List vertices in counter-clockwise order, so that faces are outwards. */
  /*  Table for local IDs of incident faces of each vertex. */
  nv = 0;
  while ((nv + 1 <= elems->size[0]) && (!(elems->data[nv] == 0))) {
    for (jj = 0; jj < 4; jj++) {
      if (opphfs->data[nv + opphfs->size[0] * jj] == 0) {
        for (loop_ub = 0; loop_ub < 3; loop_ub++) {
          isborder->data[elems->data[nv + elems->size[0] * (hf[jj + (loop_ub <<
            2)] - 1)] - 1] = TRUE;
        }
      }
    }

    nv++;
  }

  /*  Construct a vertex-to-halfface mapping. */
  nv = 0;
  while ((nv + 1 <= elems->size[0]) && (!(elems->data[nv] == 0))) {
    jj = 0;
    while ((jj + 1 < 5) && (!(elems->data[nv + elems->size[0] * jj] == 0))) {
      if (v2hf->data[elems->data[nv + elems->size[0] * jj] - 1] == 0) {
        for (loop_ub = 0; loop_ub < 3; loop_ub++) {
          if ((!isborder->data[elems->data[nv + elems->size[0] * jj] - 1]) ||
              (opphfs->data[nv + opphfs->size[0] * (v2f[jj + (loop_ub << 2)] - 1)]
               == 0)) {
            v2hf->data[elems->data[nv + elems->size[0] * jj] - 1] = (((nv + 1) <<
              3) + v2f[jj + (loop_ub << 2)]) - 1;
          }
        }
      }

      jj++;
    }

    nv++;
  }

  emxFree_boolean_T(&isborder);
  emxFree_int32_T(&b_v2hf);
}

void determine_incident_halffaces_initialize(void)
{
}

void determine_incident_halffaces_terminate(void)
{
  /* (no terminate code required) */
}

File: matlab/hi/3d/determine_incident_halffaces.m
function v2hf = determine_incident_halffaces(nv, elems, sibhfs, varargin)
%DETERMINE_INCIDENT_HALFFACES Determine an incident half-faces.
%    DETERMINE_INCIDENT_HALFFACES(ELEMS,SIBHFS,V2HF)
% Determines an incident half-faces of each vertex. It gives higher
% priorities to border faces.
%
% Input:  NV:     number of vertices
%         ELEMS:  matrix storing the element connectivity
%         SIBHFS: matrix storing the sibling half-faces
%
% Output: V2HF:   array of size equal to number of vertices.
%
% See also DETERMINE_INCIDENT_HALFEDGES, DETERMINE_INCIDENT_HALFVERTS

%#codegen -args {int32(0), coder.typeof( int32(0), [inf, inf]),
%#codegen        coder.typeof( int32(0), [inf, 6], [1,1])}

% Table for local IDs of incident faces of each vertex.
v2f_tet   = int32([2 4 1; 1 3 2; 3 1 4; 4 2 3]);
v2f_pyr   = int32([2,5,1,0; 3,2,1,0; 4,3,1,0; 5,4,1,0; 2,3,4,5]);
v2f_pri   = int32([1,3,4; 2,1,4; 3,2,4; 3,1,5; 1,2,5; 2,3,5]);
v2f_hex   = int32([2,5,1; 3,2,1; 4,3,1; 5,4,1; 6,5,2; 6,2,3; 6,3,4; 6,4,5]);

if nargin<4 || isempty(varargin{1}) || ~islogical(varargin{1})
    v2hf = zeros( nv, 1, 'int32');
else
    v2hf = struct('cid',zeros( nv, 1, 'int32'),'lfid',zeros( nv, 1, 'int8'));
end

% We use three bits for local-face ID.
if size(elems,2)==1
    if nargin<4 || isempty(varargin{1}) || ~islogical(varargin{1})
        assert( size(sibhfs,2)==1);
    else
        assert( size(sibhfs.cid,2)==1);
        assert( size(sibhfs.lfid,2)==1);
    end
    % Mixed elements
    
    isborder = determine_border_vertices_vol(nv, elems, sibhfs);
    
    offset=int32(1); offset_o=int32(1); ii = int32(1);
    while offset < size(elems,1)
        switch elems(offset)
            case {4,10}
                % Tetrahedral
                for jj=1:4
                    v = elems(offset+jj);
                    
                    if v2hf(v)==0
                        for kk=1:3
                            if ~isborder(v) || sibhfs( offset_o+v2f_tet(jj,kk))==0
                                v2hf(v) = clfids2hfid( ii, v2f_tet(jj,kk));
                            end
                        end
                    end
                end
            case {5,14}
                % Pyramid
                for jj=1:5
                    v = elems(offset+jj);
                    
                    if v2hf(v)==0
                        for kk=1:3+(jj==4)
                            if ~isborder(v) || sibhfs( offset_o+v2f_pyr(jj,kk))==0
                                v2hf(v) = clfids2hfid( ii, v2f_pyr(jj,kk));
                            end
                        end
                    end
                end
            case {6,15,18}
                % Prism
                for jj=1:6
                    v = elems(offset+jj);
                    
                    if v2hf(v)==0
                        for kk=1:3
                            if ~isborder(v) || sibhfs( offset_o+v2f_pri(jj,kk))==0
                                v2hf(v) = clfids2hfid( ii, v2f_pri(jj,kk));
                            end
                        end
                    end
                end
            case {8,20,27}
                % Hexahedral
                for jj=1:8
                    v = elems(offset+jj);
                    
                    if v2hf(v)==0
                        for kk=1:4
                            if ~isborder(v) || sibhfs( offset_o+v2f_hex(jj,kk))==0
                                v2hf(v) = clfids2hfid( ii, v2f_hex(jj,kk));
                            end
                        end
                    end
                end
            otherwise
                error('unsupported element type');
        end
        
        ii = ii + 1;
        offset = offset+elems(offset)+1;
        offset_o = offset_o + sibhfs(offset_o) + 1;
    end
else
    nvpE = int32(size(elems,2));
    
    % Table for local IDs of incident faces of each vertex.
    switch nvpE
        case {4,10}
            ncvpE = int32(4);
            v2f = v2f_tet;
        case {5,14}
            ncvpE = int32(5);
            v2f = v2f_pyr;
        case {6,15,18}
            ncvpE = int32(6);
            v2f = v2f_pri;
        case {8,20,27}
            ncvpE = int32(8);
            v2f = v2f_hex;
        otherwise
            error('Unsupported element type');
            ncvpE = int32(0); v2f = v2f_hex; %#ok<UNRCH>
    end
    
    isborder = determine_border_vertices_vol(nv, elems, sibhfs);
    
    % Construct a vertex-to-halfface mapping.
    ii = int32(1);
    while ii<=int32(size(elems,1))
        if elems(ii,1)==0; break; end
        for jj=1:ncvpE
            v = elems(ii,jj);
            if v==0; break; end
            
            if v2hf(v)==0
                for kk=1:(3+(size(v2f,2)>3 && v2f(jj,end)))
                    if ~isborder(v) || sibhfs( ii,v2f(jj,kk))==0
                        v2hf(v) = clfids2hfid( ii, v2f(jj,kk));
                    end
                end
            end
        end
        ii=ii+1;
    end
end
